Finer grain ds metrics

rafatower opened the pull request:
A quick&dirty way of getting DS detailed metrics, with user/day granularity. Took me longer than expected, because of the need to overcome/workaround the always-calculate-totals-for-org-users thing. There's room for improvement.

I leave it here, to the benefit of others and to others to decide.

juanignaciosl commented on app/helpers/data_services_metrics_helper.rb:
```diff
@@ -67,17 +67,17 @@ def get_geocoding_data(user, from, to)
 
-  def get_isolines_data(user, from, to)
-    orgname = user.organization.nil? ? nil : user.organization.name
+  def get_isolines_data(user, from, to, orgwise=true)
```
Consider adding support for `orgwise` to other helper methods in this class, such as `get_obs_snapshot_data` (aka: _ya que te pones..._).

rafatower replied:
Done, for `get_obs_snapshot_data` and `get_obs_general_data`

juanignaciosl commented on app/models/carto/user.rb:
```diff
@@ -379,13 +379,15 @@ def add_oauth(service, token)
     date_to = (options[:to] ? options[:to].to_date : Date.today)
     date_from = (options[:from] ? options[:from].to_date : last_billing_cycle)
-    get_user_geocoding_data(self, date_from, date_to)
+    orgwise = options.fetch(:orgwise, true)
```
Consider extracting a method:

```ruby
def ds_data_parameters_from_options(options)
  date_to = (options[:to] ? options[:to].to_date : Date.today)
  date_from = (options[:from] ? options[:from].to_date : last_billing_cycle)
  orgwise = options.fetch(:orgwise, true)
  return date_from, date_to, orgwise
end

rafatower replied:
Done

juanignaciosl commented on lib/tasks/metrics.rake:
```diff
@@ -43,5 +43,55 @@ namespace :cartodb do
+      from = args[:from].blank? ? nil : args[:from].to_date
+      to = args[:to].blank? ? nil : args[:to].to_date
+      default_output_file = "/tmp/ds_metrics_#{username}_#{from.strftime('%Y%m%d')}_#{to.strftime('%Y%m%d')}.csv"
+      output_file = args[:output_file].blank? ? default_output_file : args[:output_file]
```
I think that you can use `args.fetch(:output_file, default_output_file)` instead of the ternary check.

rafatower replied:
Done

juanignaciosl commented on lib/tasks/metrics.rake:
```diff
@@ -43,5 +43,55 @@ namespace :cartodb do
+      default_output_file = "/tmp/ds_metrics_#{username}_#{from.strftime('%Y%m%d')}_#{to.strftime('%Y%m%d')}.csv"
+      output_file = args[:output_file].blank? ? default_output_file : args[:output_file]
+      CSV.open(output_file, "wb") do |csv|
+        user = Carto::User.where(username: username).first
```
Better fetch the user before opening the file so in case of error it's not created.

rafatower replied:
Done

rafatower commented:
> I leave it here, to the benefit of others and to others to decide.

Actually my original plan was to leave it there as-is. In the end I addressed all the comments, pleased hound and refined it a little :slightly_smiling_face: 

@juanignaciosl if you think it does more good than harm now (as I do), can I merge it?

rafatower commented on app/helpers/data_services_metrics_helper.rb:
```diff
@@ -7,26 +7,26 @@
 module DataServicesMetricsHelper
-  def get_user_geocoding_data(user, from, to)
-    get_geocoding_data(user, from, to)
+  def get_user_geocoding_data(user, from, to, orgwise=true)
```
All honored

rafatower commented:
I tested it in staging and turns out the commit 81ac425d0f8cc97752e68884528c051a5f67a39b with `args.fetch(:output_file, default_output_file)` is wrong:
```
$ RAILS_ENV=staging bundle exec rake --trace cartodb:metrics:ds_user_metrics['rtorre-analysis','2019-01-01','2019-02-28']
** Invoke cartodb:metrics:ds_user_metrics (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute cartodb:metrics:ds_user_metrics
rake aborted!
no implicit conversion of nil into String
/usr/lib/ruby/2.2.0/csv.rb:1256:in `initialize'
/usr/lib/ruby/2.2.0/csv.rb:1256:in `open'
/usr/lib/ruby/2.2.0/csv.rb:1256:in `open'
/home/ubuntu/www/production.cartodb.com/releases/20190218140149/lib/tasks/metrics.rake:57:in `block (3 levels) in <top (required)>'
```
 So I will revert it, test it and deploy if fine.

juanignaciosl commented:
> if you think it does more good than harm now (as I do), can I merge it?

Yes, it's certainly not an historic milestone for CARTO but it's nice :-) Don't leave abandoned branches around, pass it to acceptance and merge :rocket: 

About the break, damn. I've been away from Ruby for months now, I didn't remember that rake `args` is not a normal hash U_U.

juanignaciosl commented:
@rafatower the right approach for `args` is using `with_defaults`. I didn't remember it. It's easy and neat, give it a try.
